feat: add organization bookings statistics endpoint

Devanshusharma2005 opened the pull request:
## What does this PR do?

This PR implements organization bookings statistics endpoints for the Cal.com v2 API, as requested in GitHub issue #22468. It adds two new endpoints that provide comprehensive booking statistics and counts for organizations, following the pattern established in PR #22473.

- Fixes #22468

**New Endpoints:**
- `GET /v2/organizations/{orgId}/bookings/count` - Returns booking count with filtering
- `GET /v2/organizations/{orgId}/bookings/statistics` - Returns comprehensive booking statistics

**Key Features:**
- Efficient counting using `getBookingsCount` from platform-libraries
- Support for filtering by status, userIds, eventTypeIds, date ranges
- Organization-level authentication (`ORG_ADMIN` role required)
- Platform plan validation (requires `ESSENTIALS` plan)
- Comprehensive booking metrics: created, completed, rescheduled, cancelled, pending, no-show, average rating, CSAT score

## Visual Demo (For contributors especially)

#### API Response Examples:

**Count Endpoint Response:**
```json
{
  "status": "success",
  "data": {
    "count": 156
  }
}
```

**Statistics Endpoint Response:**
```json
{
  "status": "success", 
  "data": {
    "created": 156,
    "completed": 120,
    "rescheduled": 15,
    "cancelled": 21,
    "pending": 12,
    "noShow": 8,
    "averageRating": 4.2,
    "csatScore": 85
  }
}
```

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. **N/A** - API endpoints are auto-documented via Swagger annotations.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

**Environment Setup:**
- Requires organization with `ESSENTIALS` platform plan
- User must have `ORG_ADMIN` role for the organization
- API authentication via platform API key or OAuth token

**Test Data Requirements:**
- Organization with bookings in various statuses (ACCEPTED, CANCELLED, PENDING, etc.)
- Bookings with different event types and users
- Some bookings with ratings for CSAT calculation

**Manual Testing:**
```bash
# Test count endpoint
curl -H "Authorization: Bearer YOUR_API_KEY" \
  "https://api.cal.com/v2/organizations/123/bookings/count?status=ACCEPTED,PENDING"

# Test statistics endpoint  
curl -H "Authorization: Bearer YOUR_API_KEY" \
  "https://api.cal.com/v2/organizations/123/bookings/statistics?startDate=2024-01-01&endDate=2024-12-31"
```

**Expected Behavior:**
- Count endpoint returns accurate booking counts based on filters
- Statistics endpoint returns correct calculations for all metrics
- Both endpoints respect organization boundaries and user permissions
- Status filtering works correctly with `queryParams.status || []`

## Checklist

- [x] I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked if my changes generate no new warnings

## Critical Review Areas

**⚠️ High Priority Review Items:**

1. **Statistics Calculation Logic** - Verify the `calculateAverageRating()` and `calculateCSATScore()` methods produce accurate results
2. **Performance Implications** - The statistics endpoint fetches all bookings to calculate metrics, which could be slow for large organizations
3. **Authentication Guards** - Ensure `@Roles("ORG_ADMIN")` and `@PlatformPlan("ESSENTIALS")` are properly enforced
4. **Status Filtering Fix** - Confirm `queryParams.status || []` correctly replaces the previous hardcoded empty arrays
5. **Type Safety** - Some implementations use `any[]` for bookings which may hide type issues

**🔍 Testing Validation:**
- E2E tests cover authentication, filtering, and data accuracy scenarios
- Status filtering tests validate the bug fix works correctly
- All critical CI checks (Type check, Linters, Unit Tests) are passing

**📝 Implementation Notes:**
- Uses efficient `getBookingsCount` from platform-libraries for count endpoint
- Statistics endpoint still requires fetching all bookings for metric calculations
- Follows existing Cal.com API patterns for consistency
- Includes comprehensive Swagger documentation

---

**Link to Devin run:** https://app.devin.ai/sessions/ebed3dc0af0a4fde8dda7de9f4a865b6  
**Requested by:** @Devanshusharma2005

devin-ai-integration[bot] committed:
feat: add organization bookings statistics endpoint

- Add GetOrganizationsBookingsStatisticsInput and Output types
- Implement getBookingsStatistics method in BookingsService_2024_08_13
- Add /statistics endpoint to OrganizationsBookingsController
- Export getBookingsCount from platform-libraries
- Follow pattern from PR #22473 for counting logic
- Support comprehensive booking statistics: created, completed, rescheduled, cancelled, pending, no-show, rating, CSAT

Addresses GitHub issue #22468

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

devin-ai-integration[bot] committed:
fix: resolve missing GetBookingsCountInput and GetBookingsCountOutput types

- Remove non-existent GetBookingsCountInput_2024_08_13 import
- Make GetOrganizationsBookingsCountInput extend GetBookingsInput_2024_08_13
- Create missing GetBookingsCountOutput_2024_08_13 type
- Export GetBookingsCountOutput_2024_08_13 from bookings outputs

Fixes CI type error: 'GetBookingsCountInput_2024_08_13' has no exported member

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

supalarry reviewed (commented):
Good job - have some thoughts:

devin-ai-integration[bot] committed:
perf: optimize getBookingsCount to use efficient counting instead of fetching all bookings

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

devin-ai-integration[bot] committed:
test: add comprehensive e2e tests for organization bookings count and statistics endpoints

- Add tests for /count endpoint with various filters (userIds, eventTypeIds)
- Add tests for /statistics endpoint with date range and filter validation
- Test authentication and authorization scenarios
- Validate response structure and data accuracy
- Follow existing test patterns with proper setup and cleanup

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

Devanshusharma2005 committed:
Merge branch 'main' into devin/1755747084-bookings-statistics-endpoint

supalarry reviewed (changes requested):
Review

devin-ai-integration[bot] committed:
fix: move getBookingsCount import to top of file and use platform-libraries export

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

devin-ai-integration[bot] committed:
fix: use queryParams.status for bookingListingByStatus and add e2e tests for status filtering

- Fix both getBookingsCount and getBookingsStatistics to use queryParams.status || [] instead of hardcoded empty array
- Add comprehensive e2e tests for status filtering on both count and statistics endpoints
- Remove unused imports to fix ESLint warnings

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

devin-ai-integration[bot] committed:
fix: change 'completed' to 'accepted' in statistics to match actual booking status

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

devin-ai-integration[bot] committed:
test: replace generic type checks with specific expected values for rating calculations

- Update e2e tests to check for specific values (0) instead of just verifying number type
- Based on test data setup where bookings are created without ratings via API
- Ensures rating calculation logic is actually correct, not just type-safe

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

devin-ai-integration[bot] committed:
refactor: remove duplicate GetOrganizationsBookingsCountInput class and unused import

Co-Authored-By: Devanshu Sharma <devanshusharma658@gmail.com>

Devanshusharma2005 committed:
Merge branch 'main' into devin/1755747084-bookings-statistics-endpoint

supalarry commented on apps/api/v2/src/modules/organizations/bookings/organizations-bookings.controller.ts:
```diff
@@ -57,4 +62,55 @@ export class OrganizationsBookingsController {
+    @Param("orgId", ParseIntPipe) orgId: number,
+    @GetUser() user: UserWithProfile
+  ): Promise<GetBookingsCountOutput_2024_08_13> {
+    const { userIds, ...restParams } = queryParams;
+
+    const count = await this.bookingsService.getBookingsCount(
+      { ...restParams },
+      { email: user.email, id: user.id, orgId },
+      userIds
+    );
+
+    return {
+      status: SUCCESS_STATUS,
+      data: count,
+    };
+  }
+
+  @Get("/statistics")
+  @ApiOperation({ summary: "Get organization booking statistics" })
+  @Roles("ORG_ADMIN")
+  @PlatformPlan("ESSENTIALS")
+  @HttpCode(HttpStatus.OK)
+  async getOrganizationBookingStatistics(
+    @Query() queryParams: GetOrganizationsBookingsStatisticsInput,
+    @Param("orgId", ParseIntPipe) orgId: number,
+    @GetUser() user: UserWithProfile
+  ): Promise<GetOrganizationsBookingsStatisticsOutput_2024_08_13> {
+    const { userIds, ...restParams } = queryParams;
+
+    const statistics = await this.bookingsService.getBookingsStatistics(
+      { ...restParams },
+      { email: user.email, id: user.id, orgId },
+      userIds
+    );
+
+    return {
+      status: SUCCESS_STATUS,
+      data: statistics,
+    };
+  }
```
@coderabbit-ai doesn't [organizations-bookings.controller.e2e-spec.ts](https://github.com/calcom/cal.com/pull/23235/files#diff-ce5f7be857d8dc05fc74a20ad40e7b46dd08a08c325d7be7f0cfaf9f4fca8ad4) contain all the tests needed???

supalarry commented on apps/api/v2/src/ee/bookings/2024-08-13/services/bookings.service.ts:
```diff
@@ -1097,4 +1098,89 @@ export class BookingsService_2024_08_13 {
+    const fetchedBookings = await getAllUserBookings({
+      bookingListingByStatus: queryParams.status || [],
+      skip: 0,
+      take: null,
```
can we pass `take: 0` for readability and consistecy?

supalarry commented on apps/api/v2/src/ee/bookings/2024-08-13/services/bookings.service.ts:
```diff
@@ -1097,4 +1098,89 @@ export class BookingsService_2024_08_13 {
     });
   }
+
+  async getBookingsCount(
```
Can we move the newly added functions to `BookingsStatisticsService` ? if we keep adding more and more in this service will be hard to understand whats going on.

supalarry commented on packages/platform/types/organizations/bookings/get-org-bookings-statistics.input.ts:
```diff
@@ -0,0 +1,28 @@
+
+import { GetBookingsInput_2024_08_13 } from "@calcom/platform-types";
+
+export class GetOrganizationsBookingsStatisticsInput extends GetBookingsInput_2024_08_13 {
```
We already have an identical type `GetOrganizationsBookingsInput` in `packages/platform/types/organizations/bookings/get-org-bookings.input.ts` that we can use so just reuse that.

supalarry reviewed (changes requested):
Review

supalarry commented on packages/platform/libraries/index.ts:
```diff
@@ -59,6 +59,7 @@ export { createNewUsersConnectToOrgIfExists };
 export { getAllUserBookings };
 export { getBookingInfo };
 export { handleCancelBooking };
+export { getBookingsCount } from "@calcom/trpc/server/routers/viewer/bookings/get.handler";
```
indeed - left review on it myself

supalarry commented on packages/platform/types/organizations/bookings/get-org-bookings-statistics.input.ts:
```diff
@@ -0,0 +1,28 @@
+
+import { GetBookingsInput_2024_08_13 } from "@calcom/platform-types";
+
+export class GetOrganizationsBookingsStatisticsInput extends GetBookingsInput_2024_08_13 {
+  @IsArray()
+  @IsOptional()
+  @Transform(({ value }) => {
+    if (typeof value === "string") {
+      return value.split(",").map((userId: string) => parseInt(userId));
+    }
+    if (Array.isArray(value)) {
+      return value.map((userId: string | number) => +userId);
+    }
+    return value;
+  })
+  @IsNumber({}, { each: true })
+  @ArrayMinSize(1, { message: "userIds must contain at least 1 user id" })
+  @ApiProperty({
+    type: String,
+    required: false,
+    description: "Filter bookings statistics by ids of users within your organization.",
+    example: "?userIds=100,200",
+  })
+  userIds?: number[];
```
left this in my review comment too

supalarry commented on packages/platform/types/organizations/bookings/get-org-bookings.input.ts:
```diff
@@ -26,3 +29,26 @@ export class GetOrganizationsBookingsInput extends GetBookingsInput_2024_08_13 {
+    example: "?userIds=100,200",
+  })
+  userIds?: number[];
+}
```
GetOrganizationsBookingsCountInput has been replaced with GetBookingsInput_2024_08_13

supalarry commented:
Closing the issue because we will want platform dashboard which will have org level bookings count and other statistics. However, we already have /insights for individuals and teams and maybe will have for orgs (checking in about this) and then we could re-use that code.

Secondly, there was an attempt to fix this in https://github.com/calcom/cal.com/pull/23235 but it re-uses get bookings handler with take: 0 which apparently can still result in big load on db (https://github.com/calcom/cal.com/pull/23235#discussion_r2322654773). To be sure, we need a carefully crafted query just for the count so that the database is not overloaded.
